crates/ruma-identifiers/src/macros.rs

```diff
@@ -289,3 +289,175 @@ macro_rules! opaque_identifier {
         partial_eq_string!(Box<$id>);
     };
 }
+
+macro_rules! opaque_identifier_validated {
+    (
+        $( #[doc = $docs:literal] )*
+        $vis:vis type $id:ident;
+        $val_id:ident;
+    ) => {
+        $( #[doc = $docs] )*
+        #[repr(transparent)]
+        #[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]
+        #[cfg_attr(feature = "serde", derive(serde::Serialize), serde(transparent, crate = "serde"))]
+        pub struct $id(str);
+
+        paste::paste! {
+            doc_concat! {
+                #[doc = concat!("An owned [", stringify!($id), "].")]
+                pub type [<$id Box>] = Box<$id>;
+            }
+        }
+
+        impl $id {
+
+            #[allow(clippy::transmute_ptr_to_ptr)]
+            fn from_borrowed(s: &str) -> &Self {
+                unsafe { std::mem::transmute(s) }
+            }
+
+            pub(super) fn from_owned(s: Box<str>) -> Box<Self> {
+                unsafe { Box::from_raw(Box::into_raw(s) as _) }
+            }
+
+            fn into_owned(self: Box<Self>) -> Box<str> {
+                unsafe { Box::from_raw(Box::into_raw(self) as _) }
+            }
+
+            doc_concat! {
+                #[doc = concat!("Creates a string slice from this `", stringify!($id), "`.")]
+                pub fn as_str(&self) -> &str {
+                    &self.0
+                }
+            }
+
+            doc_concat! {
+                #[doc = concat!("Creates a byte slice from this `", stringify!($id), "`.")]
+                pub fn as_bytes(&self) -> &[u8] {
+                    self.0.as_bytes()
+                }
+            }
+        }
+
+        impl std::fmt::Debug for $id {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                f.write_str(self.as_ref())
+            }
+        }
+
+        impl Clone for Box<$id> {
+            fn clone(&self) -> Self {
+                (**self).to_owned()
+            }
+        }
+
+        impl ToOwned for $id {
+            type Owned = Box<$id>;
+
+            fn to_owned(&self) -> Self::Owned {
+                Self::from_owned(self.0.to_owned().into_boxed_str())
```

Comment:
Same as all others, now :)

```suggestion
                Self::from_owned(self.0.into())
```